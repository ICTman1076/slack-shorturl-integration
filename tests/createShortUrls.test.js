const createShortUrlsFactory = require('../src/createShortUrls')

jest.mock('request-promise-native', () => (options) => {
  const unauthorizedResponse = require('./fixtures/rebrandlyResponses/unauthorized.json')
  const invalidDomainResponse = require('./fixtures/rebrandlyResponses/invalidDomain.json')
  const invalidUrlResponse = require('./fixtures/rebrandlyResponses/invalidUrl.json')
  const successResponse = require('./fixtures/rebrandlyResponses/success.json')

  const body = JSON.parse(options.body)

  if (options.headers.apikey === 'invalidApiKey') {
    const invalidApiKeyError = new Error()
    invalidApiKeyError.response = unauthorizedResponse
    return Promise.reject(invalidApiKeyError)
  } else if (options.headers.apikey === 'serverError') {
    const serverError = new Error('Some Server error happened')
    serverError.response = { statusCode: 500 }
    return Promise.reject(serverError)
  } else if (body.destination === 'invalidUrl') {
    return Promise.resolve(invalidUrlResponse)
  } else if (body.domain && body.domain.fullName === 'invalidDomain') {
    return Promise.resolve(invalidDomainResponse)
  }

  return Promise.resolve(successResponse)
})

test('It should return a function', () => {
  expect(typeof createShortUrlsFactory('someValidApiKey')).toBe('function')
})

test('It should handle invalid error because of an invalid apikey', () => {
  const csu = createShortUrlsFactory('invalidApiKey')
  expect(csu(['http://example.com'])).resolves.toMatchSnapshot()
})

test('It should handle error because of invalid url', () => {
  const csu = createShortUrlsFactory('someValidApiKey')
  expect(csu(['invalidUrl'])).resolves.toMatchSnapshot()
})

test('It should handle error because of invalid domain', () => {
  const csu = createShortUrlsFactory('someValidApiKey')
  expect(csu(['http://example.com'], 'invalidDomain')).resolves.toMatchSnapshot()
})

test('It should handle a generic server error', () => {
  const csu = createShortUrlsFactory('serverError')
  expect(csu(['http://example.com'])).resolves.toMatchSnapshot()
})

test('It should successfully create a shortUrl', () => {
  const csu = createShortUrlsFactory('someValidApiKey')
  expect(csu(['http://example.com'], 'somedomain.com', ['custom'])).resolves.toMatchSnapshot()
})
