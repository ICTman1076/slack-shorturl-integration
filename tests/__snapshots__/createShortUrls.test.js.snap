// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`It should handle a generic server error 1`] = `
Array [
  [Error: Cannot create short URL for "http://example.com": Some Server error happened ()],
]
`;

exports[`It should handle error because of invalid domain 1`] = `
Array [
  [Error: Cannot create short URL for "http://example.com": Not found (domain)],
]
`;

exports[`It should handle error because of invalid url 1`] = `
Array [
  Object {
    "code": "InvalidFormat",
    "errors": Array [
      Object {
        "code": "RequiredField",
        "message": "Cannot be empty",
        "property": "destination",
        "verbose": "destination cannot be empty",
      },
    ],
    "message": "Invalid format",
    "source": "link",
  },
]
`;

exports[`It should handle invalid error because of an invalid apikey 1`] = `
Array [
  [Error: Unauthorized. Verify that you are using a valid Rebrandly apikey],
]
`;

exports[`It should successfully create a shortUrl 1`] = `
Array [
  Object {
    "clicks": 0,
    "createdAt": "2017-05-20T15:16:07.000Z",
    "creator": Object {
      "avatarUrl": "https://s.gravatar.com/avatar/e18d2c2510d828bc95bde6817e6e9ce2?size=80&d=retro&rating=g",
      "fullName": "Luciano",
      "id": "25bbd054d27e408f8ab9e0e1dd8de112",
    },
    "destination": "https://facebook.github.io/jest/docs/expect.html",
    "domain": Object {
      "fullName": "rebrand.ly",
      "id": "8f104cc5b6ee4a4ba7897b06ac2ddcfb",
      "ref": "/domains/8f104cc5b6ee4a4ba7897b06ac2ddcfb",
    },
    "domainId": "8f104cc5b6ee4a4ba7897b06ac2ddcfb",
    "domainName": "rebrand.ly",
    "favourite": false,
    "forwardParameters": true,
    "https": true,
    "id": "1611678",
    "isPublic": false,
    "linkId": 1611678,
    "scripts": Array [],
    "shortUrl": "rebrand.ly/dxws",
    "slashtag": "dxws",
    "status": "active",
    "tags": Array [],
    "title": "Link dxws",
    "updatedAt": "2017-05-20T15:16:07.000Z",
  },
]
`;
